File: frontend/src/services/notifications.ts
import { apiClient } from "@/api/client";
import { NotificationItem } from "@/types";

export const notificationsService = {
  list: (unread_only = false) => apiClient.get<NotificationItem[]>("/api/notifications", { params: { unread_only } }).then((r) => r.data),
  markRead: (id: string) => apiClient.post<NotificationItem>(`/api/notifications/${id}/read`).then((r) => r.data),
};

File: backend/app/api/notifications.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.database.session import get_db
from app.models.notification import Notification
from app.models.user import User
from app.schemas.notification import NotificationOut
from app.security.deps import get_current_user
from app.services.notification_service import list_for_user, mark_read

router = APIRouter(prefix="/api/notifications", tags=["notifications"])


@router.get("", response_model=list[NotificationOut])
def get_notifications(unread_only: bool = False, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    return list_for_user(db, user.id, unread_only)


@router.post("/{notification_id}/read", response_model=NotificationOut)
def read_notification(notification_id: uuid.UUID, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    notification = db.get(Notification, notification_id)
    if not notification or notification.user_id != user.id:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Notification not found")
    return mark_read(db, notification)
